### PR TITLE
Use RoundTripper for authentication (and pass an http.Client to create a new Client)

### PR DESCRIPTION
**Pull request by weppos**
This implementation follows the most common best practices around using a RoundTripper for authentication, rather than passing a credential and modifying the Do to apply the credential headers.

Definining a RoundTripper is a more flexible, powerful, and elengant approach.

This change requires to bump to 1.7 (given we add the dependency to golang/x/oauth2 that uses the context package).

Closes GH-15

### CODE REVIEW

**Comment by aeden** [comment]
Ideal test for this would be a branch of https://github.com/dnsimple/dnsimple-api-examples/tree/master/golang integrated to determine the impact, if any, on API client consumers.

**Comment by weppos** [comment]
👍 I'll work on it.

File: dnsimple/authentication.go [hunk 1]
```diff
@@ -1,68 +1,52 @@
...
-	return &httpBasicCredentials{email, password}
-}
-
-func (c *httpBasicCredentials) Headers() map[string]string {
-	return map[string]string{httpHeaderAuthorization: "Basic " + c.basicAuth(c.email, c.password)}
-}
+// RoundTrip implements the RoundTripper interface.  We just add the
+// basic auth and return the RoundTripper for this transport type.
+func (t *BasicAuthTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req2 := cloneRequest(req) // per RoundTripper contract
```
Comment (jodosha) [inline dnsimple/authentication.go]: I think it's worth to link Go docs here, as future reference.

**Comment by weppos** [comment]
@aeden here's the example code
https://github.com/dnsimple/dnsimple-api-examples/pull/7

**Comment by aeden** [comment]
Looking at the usage now in the dnsimple-api-examples (thank you for that!) this feature could be introduced without needing any changes to the API callers use. For example, `dnsimple.NewOauthTokenCredentials(oauthToken)` could return the result of ` oauth2.StaticTokenSource(&oauth2.Token{AccessToken: oauthToken})` and then `dnsimple.NewClient()` could accept accept the resulting token source as its argument and internally construct the token client using the passed-in static token source and return the dnsimple client.

As the new implementation in this PR, as well as the alternative I've presented, both break existing implementations, it leaves me with the question of *why* we want to expose the internals of oauthx as a requirement for every construction of the client versus hiding that implementation inside the existing implementation? Is there a benefit I'm missing here? Are there other use cases that are needed today aside from the token?

**Pull request merged**